gno.land/pkg/sdk/vm/keeper.go

```diff
@@ -191,30 +191,43 @@ func (vm *VMKeeper) AddPackage(ctx sdk.Context, msg MsgAddPackage) error {
 	// Parse and run the files, construct *PV.
 	m2 := gno.NewMachineWithOptions(
 		gno.MachineOptions{
-			PkgPath:   "",
-			Output:    os.Stdout, // XXX
-			Store:     store,
-			Alloc:     store.GetAllocator(),
-			Context:   msgCtx,
-			MaxCycles: vm.maxCycles,
+			PkgPath:    "",
+			Output:     os.Stdout, // XXX
+			Store:      gnostore,
+			Alloc:      gnostore.GetAllocator(),
+			Context:    msgCtx,
+			MaxCycles:  vm.maxCycles,
+			VMGasMeter: ctx.GasMeter(),
 		})
-	defer m2.Release()
-	m2.RunMemPackage(memPkg, true)
+	defer func() {
+		if r := recover(); r != nil {
+			switch r.(type) {
+			case store.OutOfGasException: // panic in consumeGas()
+				panic(r)
+			default:
+				err = errors.Wrap(fmt.Errorf("%v", r), "VM addpkg panic: %v\n%s\n",
+					r, m2.String())
+				m2.Release()
+				return
+			}
+		}
+		m2.Release()
```

Comment:
> * Why do we re-panic OutOfGasexception without releasing the machine?
Spot on! In the previous version, I released the machine in vm/gas.go. Since I removed it in the new version, I forgot to put it back before the panic occurred. I've updated it now.

```
case store.OutOfGasException: // panic in consumeGas()
				panic(r)
```

It made me think that we should consider moving back to the original Machine lifecycle model, where we do not need to manually manage the machine instance lifecycle since it lives within a context on a per-request basis. The context is managed separately by the RPC. 

To prevent the instantiation of a lot of machine instances while the node is loading packages and during playback, we can use a single machine instance to perform these tasks during initialization or when restarting nodes.

In other words, fundamentally, we don't need a machine pool and to manually manage the lifecycle, which is prone to mistakes and adds unnecessary overhead to scrutinizing the code.
			

> * Why do we need to wrap the error otherwise?
> * Why are the panics within the gnovm handled here, but there are none for `addpkg`, and in fact they are unnecessary because they are correctly and much better handled in 2 different instances?!


The gas meter lifecycle begins in baseapp during CheckTx, DeliverTx, and BeginBlock, with multiple gas consumption checks before reaching the VM. store.OutOfGasException is managed at this stage.

Upon reaching the VM, panic errors and OutOfGasException are distinctively handled. Out-of-gas exceptions are deferred to baseapp, which notifies users of insufficient gas. Other panics are encapsulated with a trace for user insight and troubleshooting. 

I agree that there is room to improve on the error wrapping with trace. We aim to respond with a trace for real VM abnormal panics; however, for panics resulting from invalid user input, probably not. We can fine-tune error display when we look at improving the client UX altogether.

Reply:
https://github.com/gnolang/gno/issues/1843